### PR TITLE
fix: ensuring indexes on 'id'

### PR DESCRIPTION
**Pull request by dwinston**
source from nmdc schema and compatible non-system collections

part of #240

### CODE REVIEW

File: nmdc_runtime/api/core/metadata.py [hunk 1]
```diff
@@ -37,11 +37,14 @@
     if "items" in db_prop_spec and "$ref" in db_prop_spec["items"]
 }
 
-collection_names_with_unique_id = {
-    coll_name
-    for coll_name, class_name in collection_name_to_class_name.items()
-    if "id" in get_nmdc_jsonschema_dict()["$defs"][class_name].get("required", [])
-}
+
+@lru_cache
```
Comment (shreddd) [inline nmdc_runtime/api/core/metadata.py:41]: Do we need to worry about cache invalidation? Or would  that only happen if a new version is deployed.

File: nmdc_runtime/api/db/mongo.py [hunk 2]
```diff
@@ -73,3 +78,107 @@ def activity_collection_names(mdb: MongoDatabase) -> Set[str]:
...
+    return len(coll.distinct("id")) == total_count
+
+
+def ensure_unique_id_indexes(mdb: MongoDatabase):
+    """Ensure that any collections with an "id" field have an index on "id"."""
+    candidate_names = (
+        set(mdb.list_collection_names()) | schema_collection_names_with_id_field()
+    )
+    for collection_name in candidate_names:
+        if collection_name.startswith("system."):  # reserved by mongodb
```
Comment (shreddd) [inline nmdc_runtime/api/db/mongo.py:99]: Should we also ignore "internal" NMDC collections beginning with _

File: nmdc_runtime/api/db/mongo.py [hunk 3]
```diff
@@ -73,3 +78,107 @@ def activity_collection_names(mdb: MongoDatabase) -> Set[str]:
...
+class DeleteStatement(BaseModel):
+    q: dict
+    limit: conint(ge=0, le=1) = 1
+
+
+class OverlayDBError(Exception):
+    pass
+
+
+class OverlayDB(AbstractContextManager):
```
Comment (shreddd) [inline nmdc_runtime/api/db/mongo.py:125]: Can you add a couple of comments explaining what the OverlayDB class is for

File: nmdc_runtime/api/core/metadata.py (hunk 1 above)
Comment (dwinston) [reply nmdc_runtime/api/core/metadata.py:41]: right, we explicitly update the `nmdc-schema` PyPI package dependency, and that necessitates a fresh deployment, so `get_nmdc_jsonschema_dict()` can be cached on init for a deployment.

File: nmdc_runtime/api/db/mongo.py (hunk 2 above)
Comment (dwinston) [reply nmdc_runtime/api/db/mongo.py:99]: Interesting. I'd like to not ignore them, and lean on the semantic that a collection with at least one document, an "id" field in every document, and all "id" values different, i.e. what is checked by `all_docs_have_unique_id`, should ensure a unique index on "id".

I can be swayed to also ignore "_"-beginning collection names if you want that to approve this PR. The current behavior is inessential.

File: nmdc_runtime/api/db/mongo.py (hunk 3 above)
Comment (dwinston) [reply nmdc_runtime/api/db/mongo.py:125]: Done.

**Review by shreddd — APPROVED** [review]
OK - this makes sense. Ready to merge.

**Pull request merged**